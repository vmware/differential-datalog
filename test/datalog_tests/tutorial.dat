start;

insert Word1("Hello,"      , CategoryOther),
insert Word1("Goodbye,"    , CategoryOther),
insert Word2("World"       , CategoryOther),
insert Word2("Ruby Tuesday", CategoryOther);

insert Word1("Help me,"      , CategoryStarWars),
insert Word1("I am your"     , CategoryStarWars),
insert Word2("Obi-Wan Kenobi", CategoryStarWars),
insert Word2("father"        , CategoryStarWars);


commit;

echo Phrases:;
dump Phrases;

start;

insert Host(0, "h0", 0xaabbccdd),
insert Host(1, "h1", 0xaabbccee),
insert Host(2, "h2", 0xaaff1234),
insert Subnet(0, 0xaabb0000, 0xffff0000),
insert Subnet(1, 0xaaff0000, 0xffff0000);

commit;

echo HostInSubnet:;
dump HostInSubnet;

start;

insert School("school1", ""),
insert School("school2", ""),
insert Student(1, "Adam", "school1", 1400),
insert Student(1, "Aaron", "school2", 1500),
insert Student(1, "Alice", "school1", 1300),
insert Student(1, "Ann", "school2", 1600),

commit;

echo TopScore;
dump TopScore;

start;

insert OnlineOrder(1, "milk"),
insert OnlineOrder(1, "eggs"),
insert OnlineOrder(1, "jackfruit"),

insert OnlineOrder(2, "sursild"),
insert OnlineOrder(2, "milk"),
insert OnlineOrder(2, "eggs"),

commit dump_changes;

start;

insert StoreInventory(StoreItem{"milk", "Just your regular cow milk"}),
insert StoreInventory(StoreItem{"kumis", "Horse milk"}),
insert StoreInventory(StoreItem{"goat milk", "Another milk variety"}),

commit dump_changes;

start;

insert Logical_Switch(0),
insert Logical_Switch(1),
insert Load_Balancer(0, 0, "10.10.10.101", std.Some{"TCP"}, "lb1"),
insert Load_Balancer(0, 0, "10.10.10.102", std.None, "lb2"),
insert Load_Balancer(0, 1, "10.10.10.103", std.Some{"UDP"}, "lb3"),
insert Load_Balancer(0, 1, "10.10.10.104", std.None, "lb4");

commit;

echo Flows:;
dump Flow;

echo Flows generated using FTL:;
dump Flow1;

start;

insert Number(1),
insert Number(10),
insert Number(100),
insert Number(1000),
insert Number(10000),
insert Number(100000),
insert Number(1000000),
insert Number(10000000),
insert Number(100000000);

commit;

echo Pow2:;
dump Pow2;

start;

insert NetHost(1, NHost{IPAddr{0xaabbccdd}, MACAddr{0x112233445566}}),
insert NetHost(2, NHost{IPAddr{0xa0b0c0d0}, MACAddr{0x102030405060}});

commit;

echo NetHostString:;
dump NetHostString;

start;

insert Bytes(10,10,10,101),
insert Bytes(127,0,0,1),
insert Bytes(224,0,0,1);

commit;

echo Address:;
dump Address;

echo MCastAddress:;
dump MCastAddress;

start;

insert Endpoint(IPAddr{0xaabbccdd}, "FTP",   0),
insert Endpoint(IPAddr{0xaabbccdd}, "HTTPS", 0),
insert Endpoint(IPAddr{0xaabbccdd}, "HTTP",  0),
insert Endpoint(IPAddr{0xaabbccdd}, "HTTP",  8080);

commit;

echo EndpointString;
dump EndpointString;

echo First5;

dump First5;

start;

insert Blacklisted("170.187.204.221:80"),
insert Blacklisted("170.187.204.221:20");

commit;

echo SanitizedEndpoint:;
dump SanitizedEndpoint;


start;

insert HostAddress(1, "10.10.10.101 10.10.10.102"),
insert HostAddress(2, "192.168.0.1"),
insert HostAddress(3, "192.168.0.3 192.168.0.4 192.168.0.5 192.168.0.6");

commit;

echo HostIP:;
dump HostIP;

echo HostIPVSep:;
dump HostIPVSep;

start;
insert Library(Book{"Ernest Hemingway", "For Whom the Bell Tolls"}),
insert Library(Book{"Dav Pilkey", "For Whom the Ball Rolls"}),
insert Author("Ernest Hemingway", 1899),
insert Author("Dav Pilkey", 1966),
commit;

echo BookByAuthor;
dump BookByAuthor;

start;
insert EvensAndOdds([0,1,2,3,4,5]),
insert EvensAndOdds([1,3,5,7]),
commit;

echo Evens:;
dump Evens;

start;
insert Vector(["a", "--", "b"], "--"),
insert Vector(["--", "a", "b"], "--"),
insert Vector(["a", "b", "--"], "--"),
commit;

echo Prefix:;
dump Prefix;

start;

insert X(10),
insert X(20),
insert X(100),
insert X(150);

commit;

echo Sum:;
dump Sum;

echo Product;
dump Product;

start;

insert Price("A", "Overpriced Inc", 120),
insert Price("A", "Scrooge Ltd", 110),
insert Price("A", "Discount Corp", 100),
insert Price("B", "Overpriced Inc", 123),
insert Price("B", "Scrooge Ltd", 150),
insert Price("B", "Discount Corp", 111);

commit;

start;

delete Price("A", "Discount Corp", 100),
delete Price("B", "Discount Corp", 111),

commit;

echo BestPrice:;
dump BestPrice;

echo WorstPrice:;
dump WorstPrice;

echo BestVendor:;
dump BestVendor;

echo BestDeal:;
dump BestDeal;

start;

insert Packet(EthPacket{
    0xaabbccddeeff,
    0x112233445566,
    EthIP4{
        IP4Pkt{
            5,
            0xa1b1c1d1,
            0xa1a2a3a4,
            IPTCP {
                TCPPkt {
                    100,
                    80,
                    0
                }
            }
        }
    }
}),
insert Packet(EthPacket{
    0x112233445566,
    0xaabbccddeeff,
    EthIP6{
        IP6Pkt{
            5,
            0xa1b1c1d1,
            0xa1a2a3a4,
            IPTCP {
                TCPPkt {
                    200,
                    8080,
                    0
                }
            }
        }
    }
}),
insert Packet(EthPacket{
    0x112233445566,
    0xaabbccddeeff,
    EthIP6{
        IP6Pkt{
            5,
            0xa1b1c1d1,
            0xa1a2a3a4,
            IPUDP {
                UDPPkt {
                    300,
                    22,
                    100
                }
            }
        }
    }
});


commit;

echo TCPDstPort:;
dump TCPDstPort;

echo UDPDstPort:;
dump UDPDstPort;

echo UDPDstPort2:;
dump UDPDstPort2;

start;

insert KnownHost(0xc0a80001),
insert KnownHost(0xc0a80102),
insert KnownHost(0xc0100001),
insert KnownHost(0x0a0a0001);

commit;

echo IntranetHost:;
dump IntranetHost;

echo IntranetHost2:;
dump IntranetHost2;

echo IntranetHost3:;
dump IntranetHost3;

start;

insert Person("Alice", "USA", "student");
insert Person("Bob", "USA", "engineer");
insert Person("Eve", "UK", "student");

commit;

echo TargetAudience:;
dump TargetAudience;
