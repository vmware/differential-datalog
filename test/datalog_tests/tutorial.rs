

fn split_ip_list(x: &String) -> Vec<String> {
    x.as_str().split(" ").map(|x| x.to_string()).collect()
}

//fn string_slice(x: &String)
fn string_slice(x: &String, from: &u64, to: &u64) -> String {
    x.as_str()[(*from as usize)..(*to as usize)].to_string()
}
