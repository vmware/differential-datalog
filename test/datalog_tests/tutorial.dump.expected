Phrases:
Phrases{.phrase = "Goodbye, Ruby Tuesday"}
Phrases{.phrase = "Goodbye, World"}
Phrases{.phrase = "Goodbye, World!"}
Phrases{.phrase = "Hello, Ruby Tuesday"}
Phrases{.phrase = "Hello, World"}
Phrases{.phrase = "Hello, World!"}
Phrases{.phrase = "Help me, Obi-Wan Kenobi"}
Phrases{.phrase = "Help me, father"}
Phrases{.phrase = "I am your Obi-Wan Kenobi"}
Phrases{.phrase = "I am your father"}
HostInSubnet:
HostInSubnet{.host = 0, .subnet = 0}
HostInSubnet{.host = 1, .subnet = 0}
HostInSubnet{.host = 2, .subnet = 1}
TopScore
TopScore{.school = "school1", .top_score = 1400}
TopScore{.school = "school2", .top_score = 1600}
ItemInOrders:
ItemInOrders{.item = "eggs", .orders = [1, 2]}: +1
ItemInOrders{.item = "jackfruit", .orders = [1]}: +1
ItemInOrders{.item = "milk", .orders = [1, 2]}: +1
ItemInOrders{.item = "sursild", .orders = [2]}: +1
MilkOrders:
MilkOrders{.order = 1}: +1
MilkOrders{.order = 2}: +1
OrderFormatted:
OrderFormatted{.order = "order: 1, item: eggs"}: +1
OrderFormatted{.order = "order: 1, item: jackfruit"}: +1
OrderFormatted{.order = "order: 1, item: milk"}: +1
OrderFormatted{.order = "order: 2, item: eggs"}: +1
OrderFormatted{.order = "order: 2, item: milk"}: +1
OrderFormatted{.order = "order: 2, item: sursild"}: +1
InventoryItemName:
InventoryItemName{.name = "goat milk"}: +1
InventoryItemName{.name = "kumis"}: +1
InventoryItemName{.name = "milk"}: +1
Flows:
Flow{.lr = 0, .stage = LS_IN_PRE_LB{}, .prio = 100, .matchStr = "ip4.dst == 10.10.10.101", .actionStr = "{ reg0[0] = 1; next; }"}
Flow{.lr = 0, .stage = LS_OUT_PRE_LB{}, .prio = 100, .matchStr = "ip4", .actionStr = "{ reg0[0] = 1; next; }"}
Flow{.lr = 1, .stage = LS_IN_PRE_LB{}, .prio = 100, .matchStr = "ip4.dst == 10.10.10.103", .actionStr = "{ reg0[0] = 1; next; }"}
Flow{.lr = 1, .stage = LS_OUT_PRE_LB{}, .prio = 100, .matchStr = "ip4", .actionStr = "{ reg0[0] = 1; next; }"}
Flows generated using FTL:
Flow1{.lr = 0, .stage = LS_IN_PRE_LB{}, .prio = 100, .matchStr = "ip4.dst == 10.10.10.101", .actionStr = "{ reg0[0] = 1; next; }"}
Flow1{.lr = 0, .stage = LS_OUT_PRE_LB{}, .prio = 100, .matchStr = "ip4", .actionStr = "{ reg0[0] = 1; next; }"}
Flow1{.lr = 1, .stage = LS_IN_PRE_LB{}, .prio = 100, .matchStr = "ip4.dst == 10.10.10.103", .actionStr = "{ reg0[0] = 1; next; }"}
Flow1{.lr = 1, .stage = LS_OUT_PRE_LB{}, .prio = 100, .matchStr = "ip4", .actionStr = "{ reg0[0] = 1; next; }"}
Pow2:
Pow2{.p = "The square of 1 is 1"}
Pow2{.p = "The square of 10 is 100"}
Pow2{.p = "The square of 100 is 10000"}
Pow2{.p = "The square of 1000 is 1000000"}
Pow2{.p = "The square of 10000 is 100000000"}
Pow2{.p = "The square of 100000 is 10000000000"}
Pow2{.p = "The square of 1000000 is 1000000000000"}
Pow2{.p = "The square of 10000000 is 100000000000000"}
Pow2{.p = "The square of 100000000 is 10000000000000000"}
NetHostString:
NetHostString{.id = 1, .s = "Host: IP=170.187.204.221, MAC=11:22:33:44:55:66"}
NetHostString{.id = 2, .s = "Host: IP=160.176.192.208, MAC=10:20:30:40:50:60"}
Address:
Address{.addr = ip_addr_t{.addr = 168430181}}
Address{.addr = ip_addr_t{.addr = 2130706433}}
Address{.addr = ip_addr_t{.addr = 3758096385}}
MCastAddress:
MCastAddress{.addr = ip_addr_t{.addr = 3758096385}}
EndpointString
EndpointString{.s = "170.187.204.221:20"}
EndpointString{.s = "170.187.204.221:443"}
EndpointString{.s = "170.187.204.221:80"}
EndpointString{.s = "170.187.204.221:8080"}
First5
First5{.str = "Goodb"}
First5{.str = "Hello"}
First5{.str = "Help "}
First5{.str = "I am "}
SanitizedEndpoint:
SanitizedEndpoint{.ep = "170.187.204.221:443"}
SanitizedEndpoint{.ep = "170.187.204.221:8080"}
HostIP:
HostIP{.host = 1, .addr = "10.10.10.101"}
HostIP{.host = 1, .addr = "10.10.10.102"}
HostIP{.host = 2, .addr = "192.168.0.1"}
HostIP{.host = 3, .addr = "192.168.0.3"}
HostIP{.host = 3, .addr = "192.168.0.4"}
HostIP{.host = 3, .addr = "192.168.0.5"}
HostIP{.host = 3, .addr = "192.168.0.6"}
HostIPVSep:
HostIPVSep{.host = 1, .addrs = "10.10.10.101\n10.10.10.102\n"}
HostIPVSep{.host = 2, .addrs = "192.168.0.1\n"}
HostIPVSep{.host = 3, .addrs = "192.168.0.3\n192.168.0.4\n192.168.0.5\n192.168.0.6\n"}
BookByAuthor
BookByAuthor{.book = Book{.author = "Dav Pilkey", .title = "For Whom the Ball Rolls"}, .author = Author{.name = "Dav Pilkey", .born = 1966}}
BookByAuthor{.book = Book{.author = "Ernest Hemingway", .title = "For Whom the Bell Tolls"}, .author = Author{.name = "Ernest Hemingway", .born = 1899}}
Evens:
Evens{.evens_and_odds = [0, 1, 2, 3, 4, 5], .evens = [0, 2, 4]}
Evens{.evens_and_odds = [1, 3, 5, 7], .evens = []}
Prefix:
Prefix{.vec = []}
Prefix{.vec = ["a"]}
Prefix{.vec = ["a", "b"]}
Sum:
Sum{.x = 10, .y = 10, .sum = 20}
Sum{.x = 10, .y = 20, .sum = 30}
Sum{.x = 10, .y = 100, .sum = 110}
Sum{.x = 10, .y = 150, .sum = 160}
Sum{.x = 20, .y = 10, .sum = 30}
Sum{.x = 20, .y = 20, .sum = 40}
Sum{.x = 20, .y = 100, .sum = 120}
Sum{.x = 20, .y = 150, .sum = 170}
Sum{.x = 100, .y = 10, .sum = 110}
Sum{.x = 100, .y = 20, .sum = 120}
Sum{.x = 100, .y = 100, .sum = 200}
Sum{.x = 100, .y = 150, .sum = 250}
Sum{.x = 150, .y = 10, .sum = 160}
Sum{.x = 150, .y = 20, .sum = 170}
Sum{.x = 150, .y = 100, .sum = 250}
Sum{.x = 150, .y = 150, .sum = 300}
Product
Product{.x = 10, .y = 10, .prod = 100}
Product{.x = 10, .y = 20, .prod = 200}
Product{.x = 10, .y = 100, .prod = 1000}
Product{.x = 10, .y = 150, .prod = 1500}
Product{.x = 20, .y = 10, .prod = 200}
Product{.x = 20, .y = 20, .prod = 400}
Product{.x = 20, .y = 100, .prod = 2000}
Product{.x = 20, .y = 150, .prod = 3000}
Product{.x = 100, .y = 10, .prod = 1000}
Product{.x = 100, .y = 20, .prod = 2000}
Product{.x = 100, .y = 100, .prod = 10000}
Product{.x = 100, .y = 150, .prod = 15000}
Product{.x = 150, .y = 10, .prod = 1500}
Product{.x = 150, .y = 20, .prod = 3000}
Product{.x = 150, .y = 100, .prod = 15000}
Product{.x = 150, .y = 150, .prod = 22500}
BestPrice:
BestPrice{.item = "A", .price = 110}
BestPrice{.item = "B", .price = 123}
WorstPrice:
WorstPrice{.item = "A", .price = 120}
WorstPrice{.item = "B", .price = 150}
BestVendor:
BestVendor{.item = "A", .vendor = "Scrooge Ltd", .price = 110}
BestVendor{.item = "B", .vendor = "Overpriced Inc", .price = 123}
BestDeal:
BestDeal{.best = "Best deal for A: Scrooge Ltd, $110"}
BestDeal{.best = "Best deal for B: Overpriced Inc, $123"}
TCPDstPort:
TCPDstPort{.port = 80}
TCPDstPort{.port = 8080}
UDPDstPort:
UDPDstPort{.port = 22}
UDPDstPort2:
UDPDstPort2{.port = 22}
IntranetHost:
IntranetHost{.addr = 3232235521}
IntranetHost{.addr = 3232235778}
IntranetHost2:
IntranetHost2{.addr = 3232235521}
IntranetHost2{.addr = 3232235778}
IntranetHost3:
IntranetHost3{.addr = 3232235521}
IntranetHost3{.addr = 3232235778}
TargetAudience:
Person{.name = "Alice", .nationality = "USA", .occupation = "student"}
